server/src/main/java/keywhiz/service/providers/ClientAuthenticator.java

```diff
@@ -0,0 +1,172 @@
+    } else {
+      logger.warn("Client {} authenticated but disabled via DB", client);
+      return Optional.empty();
+    }
+  }
+
+  private Optional<Client> handleMissingClient(boolean createMissingClient, String name,
+      String spiffeId) {
+    if (createMissingClient && !name.isEmpty()) {
+      /*
+       * If a client is seen for the first time, authenticated by certificate, and has no DB entry,
+       * then a DB entry is created here. The client can be disabled in the future by flipping the
+       * 'enabled' field.
+       */
+      long clientId = clientDAOReadWrite.createClient(name, "automatic",
+          "Client created automatically from valid certificate authentication", spiffeId);
+      return clientDAOReadWrite.getClientById(clientId);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  static Optional<String> getClientName(Principal principal) {
+    X500Name name = new X500Name(principal.getName());
+    RDN[] rdns = name.getRDNs(BCStyle.CN);
+    if (rdns.length == 0) {
+      logger.warn("Certificate does not contain CN=xxx,...: {}", principal.getName());
+      return Optional.empty();
+    }
+    return Optional.of(IETFUtils.valueToString(rdns[0].getFirst().getValue()));
+  }
+
+  static Optional<String> getSpiffeId(Principal principal) {
+    if (!(principal instanceof CertificatePrincipal)) {
+      // A SPIFFE ID can only be parsed from a principal with a certificate
+      return Optional.empty();
+    }
+
+    // This chain is either from the XFCC header's "Cert" field, which includes only the
+    // client certificate rather than the chain, or from the CertificateSecurityContext
+    // configured by Keywhiz' ClientCertificateFilter, which sets it based on
+    // X509Certificate[] chain =
+    //        (X509Certificate[]) context.getProperty("javax.servlet.request.X509Certificate");
+    // which appears to place the leaf as the zero-index entry in the chain.
+    X509Certificate cert = ((CertificatePrincipal) principal).getCertificateChain()
+        .get(0);
+    return getSpiffeIdFromCertificate(cert);
+  }
+
+  static Optional<String> getSpiffeIdFromCertificate(X509Certificate cert) {
+    Collection<List<?>> sans;
+    try {
+      sans = cert.getSubjectAlternativeNames();
+    } catch (CertificateParsingException e) {
+      logger.warn("Unable to parse SANs from principal", e);
+      return Optional.empty();
+    }
+
+    if (sans == null || sans.isEmpty()) {
+      return Optional.empty();
+    }
+
+    // The sub-lists returned by getSubjectAlternativeNames have an integer for the first
+    // entry, representing a field, and the value as a string as the second entry.
+    List<String> spiffeUriNames = sans.stream()
+        .filter(sanPair -> sanPair.get(0).equals(URINAME_SAN))
+        .map(sanPair -> (String) sanPair.get(1))
+        .filter(uri -> uri.startsWith(SPIFFE_SCHEME))
+        .collect(Collectors.toUnmodifiableList());
+
+    if (spiffeUriNames.size() > 1) {
```

Comment:
good call on limiting to only one SPIFFE ID

Reply:
I think exiting if more than one is present is a good call, it breaks with the standard so probably best to not make assumptions.